how_to_solve_a_problem.md

```diff
@@ -0,0 +1,90 @@
+
+## Finding the cause
+
+Understanding the cause of the bug is the first step and the most important one. Only in very rare cases where a serious bug happens in production and we want to first patch the bug with some quick fix, we will start by working around the problem. But we'll always circle back at finding the cause.
+
+It's better to reflect on the term "the cause". Are you sure the cause you are seeing is the "root cause"?
+Systems are always convoluted, and miss-identification of the real cause, will result with the bug simply appearing in different places.
+
+Assumptions are good but need to be taken with a pinch of salt. You shouldn't dive in too deeply just based on your assumption without verifying every step. For example, don't waste time on debugging line by line a certain module before you have asserted that it's indeed the one to blame.
+
+The first step for diving into a bug is not using a scalpel, but rather a D9. This means that the first steps don't require slow and tedious examniation of lots of code, but rather big and radical changes. Examples:
+
+| Problem | Solution |
+| -- | -- |
+| You think a bug is a result of some module. You are not sure exactly which line | Completely delete the file, and confirm there was indeed a change |
+| You are not sure if a wrong output of the screen is the result of wrong code in the theme |  Change the theme to another one, thus completely remove the theme from the equation|
+| Your new shiny code doesn't work |  Check if it is even invoked. Again, no need to have a very delicate debugging session. Instead trace the places the code is being called and figure out if is called as you assume |
+| Something that used to work stopped working | You are in luck, because you can try and find the working instance (e.g. go back in Git's history). If you are able to find a working example, the next step is to pin point the exact time in history (i.e. the git commit) that broke things, and then you'll have a smaller diff to investigate |
+
+As you can see the gist of the above examples is to work out as quickly as possible in getting into right area. From there start pin pointing the problem's cause - there is always *one specific line* that causes the havoc.
+
+After each of those above solutions we asked ourself, did it change anything?
+If yes, we knew we were in the right area, and kept zooming in.
+If not, we knew we were barking at the wrong tree. We kept using our D9 go over other parts of the code until we saw a change.
+
+![Caterpillar D9](images/solving_problem/image1.jpg)
+
+### Time boxing
+
+Like any task you must be disciplined and time box your efforts. Sometimes the _root cause_ can be buried very deep inside a legacy or a complex system. For example, trying to find the root cause of Drupal's from API can be very challenging. There are times it's needed, but there are cases a work around would be easier.
```

Comment:
I assume you will talk about time boxing in another area? I sense that there is a need for some high-level guidance to determine the time box. Not asking for a specific number. I have no clue how much time (even a range) would be appropriate to investigate a dev bug (before escalating)?